src/java/main/org/apache/zookeeper/server/NettyServerCnxn.java

```diff
@@ -100,11 +101,15 @@ public void close() {
                         + Long.toHexString(sessionId));
             }
 
-            synchronized (factory.ipMap) {
-                Set<NettyServerCnxn> s =
-                    factory.ipMap.get(((InetSocketAddress)channel
-                            .getRemoteAddress()).getAddress());
-                s.remove(this);
+            InetAddress address = ((InetSocketAddress) channel.getRemoteAddress()).getAddress();
+            Set<NettyServerCnxn> s = factory.ipMap.get(address);
+            if (s != null) {
+                synchronized (s) {
```

Comment:
Yeah... in the context of this call I think it is not necessary to synchronize here if -- as you said -- we use a synchronized set (**currently it's not**). I would also change `ipMap` to be a `ConcurrentMap` but unsure if it's worth the effort, tbh. :thinking: thoughts?